### PR TITLE
Preliminary version of libDolphin API.

### PR DESCRIPTION
**Pull request by TechNick6425**
A simple version of libDolphin
(https://docs.google.com/document/d/1KFOlRlp0a7gpgEJXaw0PGwaoNird1uqO4RZtcmmiA9c/edit)
written as a library (added to solution) and a modification of the main
emulator. A simple input handler was included, allowing other processes
to submit input to Dolphin. Also allows reading RAM values from current
game.

Restrictions:
- Only Windows (at the moment)
- It builds, but hasn't been tested

### CODE REVIEW

**Comment by BhaaLseN** [comment]
Sorry, but the code style is all over the place. Can you please take a look at [Contributing.md](https://github.com/dolphin-emu/dolphin/blob/master/Contributing.md) for reference?

Other than that, I don't think the gist of libDolphin was to provide a Named Pipe for communication, but to have a well-defined interface between the frontend (which might be WX, Qt, command line, Android, FifoCI or whatever it may be) and the backend which does the heavy lifting instead of calling various functions in the right places from main().
While its something, I don't think it goes in the right direction. You might want to join us on IRC to discuss further things if you intend to look into this further.

**Comment by lioncash** [comment]
Agreeing with @BhaaLseN in terms of styling. Also, why would you not test the very thing you created before attempting to get it merged?

File: Source/libDolphin/Release/libDolphin.log [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+         Release\Stdafx.obj
+         "Release\.NETFramework,Version=v4.5.AssemblyAttributes.obj"
+     1>LINK : warning LNK4068: /MACHINE not specified; defaulting to X86
+     1>.NETFramework,Version=v4.5.AssemblyAttributes.obj : warning LNK4221: This object file does not define any previously undefined public symbols, so it will not be used by any link operation that consumes this library
+     1>AssemblyInfo.obj : warning LNK4221: This object file does not define any previously undefined public symbols, so it will not be used by any link operation that consumes this library
+         libDolphin.vcxproj -> C:\Users\Nicholas\Documents\GitHub\dolphin\Source\Release\libDolphin.lib
+     1>Done Building Project "C:\Users\Nicholas\Documents\GitHub\dolphin\Source\libDolphin\libDolphin.vcxproj" (Build target(s)).
+
+Build succeeded.
+
```
Comment (lioncash) [inline Source/libDolphin/Release/libDolphin.log]: This file is entirely unnecessary

File: Source/libDolphin/Release/libDolphin.Build.CppClean.log [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+c:\users\nicholas\documents\github\dolphin\source\libdolphin\release\libdolphin.tlog\link.read.1.tlog
+c:\users\nicholas\documents\github\dolphin\source\libdolphin\release\libdolphin.tlog\link.write.1.tlog
+c:\users\nicholas\documents\github\dolphin\source\libdolphin\release\libdolphin.tlog\metagen.read.1.tlog
+c:\users\nicholas\documents\github\dolphin\source\libdolphin\release\libdolphin.tlog\rc.command.1.tlog
+c:\users\nicholas\documents\github\dolphin\source\libdolphin\release\libdolphin.tlog\rc.read.1.tlog
+c:\users\nicholas\documents\github\dolphin\source\libdolphin\release\libdolphin.tlog\rc.write.1.tlog
+c:\users\nicholas\documents\github\dolphin\source\libdolphin\release\libdolphin.tlog\system.data.dll.metagen
+c:\users\nicholas\documents\github\dolphin\source\libdolphin\release\libdolphin.tlog\system.dll.metagen
+c:\users\nicholas\documents\github\dolphin\source\libdolphin\release\libdolphin.tlog\system.xml.dll.metagen
+c:\users\nicholas\documents\github\dolphin\source\libdolphin\release\libdolphin.tlog\tzres.dll.bi
```
Comment (lioncash) [inline Source/libDolphin/Release/libDolphin.Build.CppClean.log]: Unnecessary file.

File: Source/libDolphin/libDolphin.cpp [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+
+	u64 input1;
+	u64 input2;
+
+	// Create a u64 to represent input.
+	// Uses the hardcoded values of GCPadStatus, and assumes no error.
+	input1 = ((u64)pad.button << 48) + ((u64)pad.stickX << 40) + ((u64)pad.stickY << 32) + ((u64)pad.substickX << 24) + ((u64)pad.substickY << 16) + ((u64)pad.triggerLeft << 8) + ((u64)pad.triggerRight);
+	input2 = ((u64)pad.analogA << 56) + ((u64)pad.analogB << 48) + 0x80807f80807f;
+
+	input[0] = input1;
```
Comment (lioncash) [inline Source/libDolphin/libDolphin.cpp]: This is entirely undefined behavior. input is a dangling pointer.

File: Source/libDolphin/libDolphin.cpp [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+			NULL
+		);
+
+		if ((bytes != 2) || (buffer[0] != 0x1A)) {
+			return -2;
+		}
+	}
+}
+
+u64* Dolphin::InterpretInput(struct GCPadStatus pad) {
```
Comment (lioncash) [inline Source/libDolphin/libDolphin.cpp]: struct is not necessary in C++, you can simply declare it as GCPadStatus, and it should be passed by const reference.

File: Source/libDolphin/libDolphin.cpp [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+	return input;
+}
+
+void Dolphin::QueueInput(struct GCPadStatus pad) {
+	m_current_pad = pad;
+}
+
+u8 Dolphin::Read8(u32 address) {
+	char *buffer;
+	buffer[0] = 0x01;
```
Comment (lioncash) [inline Source/libDolphin/libDolphin.cpp]: buffer does not point to anything, this is incorrect.

File: Source/libDolphin/libDolphin.cpp [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+	if ((bytes != 3) || (rbuffer[0] != 0x11)) {
+		return 0;
+	}
+
+	return (u8)rbuffer[1];
+}
+
+u16 Dolphin::Read16(u32 address) {
+	char *buffer;
+	buffer[0] = 0x02;
```
Comment (lioncash) [inline Source/libDolphin/libDolphin.cpp]: again, this is incorrect.

File: Source/libDolphin/libDolphin.cpp [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+	if ((bytes != 4) || (rbuffer[0] != 0x12)) {
+		return 0;
+	}
+
+	return ((u16)rbuffer[1] << 8) + ((u16)rbuffer[2]);
+}
+
+u32 Dolphin::Read32(u32 address) {
+	char *buffer;
+	buffer[0] = 0x03;
```
Comment (lioncash) [inline Source/libDolphin/libDolphin.cpp]: and again

File: Source/libDolphin/libDolphin.cpp [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+	if ((bytes != 6) || (rbuffer[0] != 0x13)) {
+		return 0;
+	}
+
+	return ((u16)rbuffer[1] << 24) + ((u16)rbuffer[2] << 16) + ((u16)rbuffer[3] << 8) + ((u16)rbuffer[4]);
+}
+
+u64 Dolphin::Read64(u32 address) {
+	char *buffer;
+	buffer[0] = 0x04;
```
Comment (lioncash) [inline Source/libDolphin/libDolphin.cpp]: yet again

File: Source/libDolphin/libDolphin.h [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+typedef int8_t  s8;
+typedef int16_t s16;
+typedef int32_t s32;
+typedef int64_t s64;
+
+#endif
+// For using Windows lock code
+#define TCHAR char
+#define LONG int
```
Comment (lioncash) [inline Source/libDolphin/libDolphin.h]: These types are already defined in CommonTypes. Simply include that header.

File: Source/libDolphin/libDolphin.h [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+	u32    Read32(u32);
+	u64    Read64(u32);
+
+	void   Write8(u32, u8);
+	void   Write16(u32, u16);
+	void   Write32(u32, u32);
+	void   Write64(u32, u64);
+
+	void   QueueInput(struct GCPadStatus);
+	u64*   InterpretInput(struct GCPadStatus);
```
Comment (lioncash) [inline Source/libDolphin/libDolphin.h]: const reference. Also struct isn't necessary.

File: Source/Core/Core/LibDolphin.cpp [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+				input[8] = ibuffer[8];
+				input[9] = ibuffer[9];
+				input[10] = ibuffer[10];
+				input[11] = ibuffer[11];
+				input[12] = ibuffer[12];
+				input[13] = ibuffer[13];
+				input[14] = ibuffer[14];
+				input[15] = ibuffer[15];
+
+				struct GCPadStatus pad;
```
Comment (lioncash) [inline Source/Core/Core/LibDolphin.cpp]: `struct` is not necessary

File: Source/Core/Core/LibDolphin.cpp [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+				emu_pad->SetEmulatedInput(pad);
+
+				break;
+			}
+
+				// Frame advance
+			case 0x0A:
+			{
+				// I know it's bad practice, but we have to break from the while loop.
+				goto finish;
```
Comment (lioncash) [inline Source/Core/Core/LibDolphin.cpp]: I would consider a different design if this if this is what's required to end the loop.

**Comment by lioncash** [comment]
I would strongly urge you to test your API, there's quite a bit of problems that are evident overall.

**Comment by comex** [comment]
You don't need to be so negative...

**Comment by lioncash** [comment]
Looking back at this, indeed I shouldn't have been. I apologize, @TechNick6425, if it seemed like I was being too harsh.

**Comment by TechNick6425** [comment]
Thanks for the feedback. I've been away for a while, so I just read this. I'll probably do some more work with this, including fixing the code style, and doing more testing. I'll probably be back in a while :)

**Comment by TechNick6425** [comment]
Hi, I'm back.

I've been having some problems with the project, and I have to drop it to move on to other things. I've left the code in my fork of the repository, so if anyone wants to try and fix it, it will be there.

Thanks for all the help.

**Pull request closed without merging**